src/coprocessor/dag/expr/arithmetic.rs

```diff
@@ -0,0 +1,407 @@
+// Copyright 2017 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+use std::{f64, i64, u64};
+use std::ops::Neg;
+use std::num::Wrapping;
+use std::borrow::Cow;
+use coprocessor::codec::{mysql, Datum};
+use coprocessor::codec::mysql::Decimal;
+use super::{Error, FnCall, Result, StatementContext};
+
+impl FnCall {
+    pub fn plus_real(&self, ctx: &StatementContext, row: &[Datum]) -> Result<Option<f64>> {
+        let lhs = try_opt!(self.children[0].eval_real(ctx, row));
+        let rhs = try_opt!(self.children[1].eval_real(ctx, row));
+        let res = lhs + rhs;
+        if !res.is_finite() {
+            return Err(Error::Overflow);
+        }
+        Ok(Some(res))
+    }
+
+    pub fn plus_decimal<'a, 'b: 'a>(
+        &'b self,
+        ctx: &StatementContext,
+        row: &'a [Datum],
+    ) -> Result<Option<Cow<'a, Decimal>>> {
+        let lhs = try_opt!(self.children[0].eval_decimal(ctx, row));
+        let rhs = try_opt!(self.children[1].eval_decimal(ctx, row));
+        let result: Result<Decimal> = (lhs.as_ref() + rhs.as_ref()).into();
```

Comment:
lhs and rhs are both `Cow<Decimal>`. We can't use `lhs.add(&rhs)` directly. `as_ref` is still needed.